File: app/src/main/java/com/example/cooplas/activities/profile/EditProfileScreen.java
package com.example.cooplas.activities.profile;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.webkit.MimeTypeMap;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.example.cooplas.R;
import com.example.cooplas.events.VerifyEvent;
import com.example.cooplas.utils.CircleTransform;
import com.example.cooplas.utils.retrofitJava.APIClient;
import com.example.cooplas.utils.retrofitJava.APIInterface;
import com.github.dhaval2404.imagepicker.ImagePicker;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.jobesk.gong.utils.FunctionsKt;
import com.kaopiz.kprogresshud.KProgressHUD;
import com.karumi.dexter.Dexter;
import com.karumi.dexter.MultiplePermissionsReport;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.multi.MultiplePermissionsListener;
import com.squareup.picasso.Picasso;

import org.greenrobot.eventbus.EventBus;

import java.io.File;
import java.util.List;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class EditProfileScreen extends AppCompatActivity {


    private EditText et_first_name, et_last_name, et_user_name;
    private String firstName, lastName, userName;
    private ImageView iv_camera, iv_profile;
    private File imageFile;

    private boolean imageSelected = false;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_profile_screen);


        iv_profile = findViewById(R.id.iv_profile);

        et_first_name = findViewById(R.id.et_first_name);
        et_last_name = findViewById(R.id.et_last_name);
        et_user_name = findViewById(R.id.et_user_name);

        et_first_name.setText(FunctionsKt.getFirstName(getApplicationContext()));
        et_last_name.setText(FunctionsKt.getLastName(getApplicationContext()));
        et_user_name.setText(FunctionsKt.getUserName(getApplicationContext()));

        TextView tv_female = findViewById(R.id.tv_female);
        TextView tv_male = findViewById(R.id.tv_male);


        Picasso.get().load(FunctionsKt.getImage(getApplicationContext())).transform(new CircleTransform()).fit().centerCrop().into(iv_profile);

        RelativeLayout rl_female = findViewById(R.id.rl_female);
        RelativeLayout rl_male = findViewById(R.id.rl_male);


        iv_camera = findViewById(R.id.iv_camera);

        String gender = FunctionsKt.getGener(getApplicationContext());
        if (gender.equalsIgnoreCase("male")) {
            rl_male.setBackground(getResources().getDrawable(R.drawable.round_orange));
            rl_female.setBackground(getResources().getDrawable(R.drawable.round_grey));

            tv_male.setTextColor(getResources().getColor(R.color.black));
            tv_female.setTextColor(getResources().getColor(R.color.black));
        } else {
            rl_female.setBackground(getResources().getDrawable(R.drawable.round_orange));
            rl_male.setBackground(getResources().getDrawable(R.drawable.round_grey));
            tv_female.setTextColor(getResources().getColor(R.color.black));
            tv_male.setTextColor(getResources().getColor(R.color.black));
        }


        rl_female.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                rl_female.setBackground(getResources().getDrawable(R.drawable.round_orange));
                rl_male.setBackground(getResources().getDrawable(R.drawable.round_grey));
                tv_female.setTextColor(getResources().getColor(R.color.black));
                tv_male.setTextColor(getResources().getColor(R.color.black));

            }
        });
        rl_male.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                rl_male.setBackground(getResources().getDrawable(R.drawable.round_orange));
                rl_female.setBackground(getResources().getDrawable(R.drawable.round_grey));

                tv_male.setTextColor(getResources().getColor(R.color.black));
                tv_female.setTextColor(getResources().getColor(R.color.black));

            }
        });

        RelativeLayout rl_save = findViewById(R.id.rl_save);
        rl_save.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                validateData();

            }
        });


        iv_profile.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                iv_camera.callOnClick();
            }
        });
        iv_camera.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Dexter.withContext(EditProfileScreen.this)
                        .withPermissions(
                                Manifest.permission.CAMERA,
                                Manifest.permission.WRITE_EXTERNAL_STORAGE
                        ).withListener(new MultiplePermissionsListener() {
                    @Override
                    public void onPermissionsChecked(MultiplePermissionsReport report) {

                        if (report.areAllPermissionsGranted()) {

                            ImagePicker.Companion.with(EditProfileScreen.this)
//                                    .crop()	    			//Crop image(Optional), Check Customization for more option
                                    .compress(1024)            //Final image size will be less than 1 MB(Optional)
                                    .maxResultSize(600, 600)    //Final image resolution will be less than 1080 x 1080(Optional)
                                    .start();
                        }

                    }

                    @Override
                    public void onPermissionRationaleShouldBeShown(List<PermissionRequest> permissions, PermissionToken token) {

                        token.continuePermissionRequest();


                    }
                }).check();
            }
        });


    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == Activity.RESULT_OK) {
            //Image Uri will not be null for RESULT_OK


            Uri fileUris = data.getData();
            iv_profile.setImageURI(fileUris);
            imageFile = new File(fileUris.getPath());

            imageSelected = true;
            Log.d("TAG", "onActivityResult: ");

        } else if (resultCode == ImagePicker.RESULT_ERROR) {
//            Toast.makeText(this, ImagePicker.getError(data), Toast.LENGTH_SHORT).show()
        } else {
            Toast.makeText(this, "Task Cancelled", Toast.LENGTH_SHORT).show();
        }


    }


    private void validateData() {

        firstName = et_first_name.getText().toString().trim();
        lastName = et_last_name.getText().toString().trim();
        userName = et_user_name.getText().toString().trim();


        if (firstName.isEmpty()) {
            Toast.makeText(this, "Please Enter First Name", Toast.LENGTH_SHORT).show();
            return;
        }
        if (lastName.isEmpty()) {
            Toast.makeText(this, "Please Enter Last Name", Toast.LENGTH_SHORT).show();
            return;
        }
        if (userName.isEmpty()) {
            Toast.makeText(this, "Please Enter User Name", Toast.LENGTH_SHORT).show();
            return;
        }

        updateProfile();
    }

    private void updateProfile() {

        KProgressHUD progressHUD = KProgressHUD.create(EditProfileScreen.this).show();
        progressHUD.show();
        String accessToken = FunctionsKt.getAccessToken(EditProfileScreen.this);
        APIInterface apiInterface = APIClient.getClient(getApplicationContext()).create(APIInterface.class);


        MultipartBody.Builder builder = new MultipartBody.Builder();
        builder.setType(MultipartBody.FORM);


        builder.addFormDataPart("first_name", firstName);
        builder.addFormDataPart("last_name", lastName);
        builder.addFormDataPart("username", userName);


        if (imageSelected == true) {


            String mimeType = getMimeType(imageFile.getPath());
            File file = new File(imageFile.getPath());
            RequestBody requestImage = RequestBody.create(MediaType.parse(mimeType), file);
//            builder.addFormDataPart("files[]", file.getName(), RequestBody.create(MediaType.parse("multipart/form-data"), file));
            builder.addFormDataPart("profile_pic", file.getName(), RequestBody.create(MediaType.parse(mimeType), file));

        }


        MultipartBody requestBody = builder.build();
        Call<JsonObject> call = apiInterface.updateProfile("Bearer " + accessToken, requestBody);
        call.enqueue(new Callback<JsonObject>() {
            @Override
            public void onResponse(Call<JsonObject> call, Response<JsonObject> response) {
                Log.d("postUpload", "" + new Gson().toJson(response.body()));
                progressHUD.dismiss();
                if (response.isSuccessful()) {
                    Toast.makeText(EditProfileScreen.this, "Your Profile Is Updated Successfully", Toast.LENGTH_SHORT).show();
                }

                EventBus.getDefault().post(new VerifyEvent());
            }

            @Override
            public void onFailure(Call<JsonObject> call, Throwable t) {
                Log.d("onFailure", t + "");
                call.cancel();
                progressHUD.dismiss();
            }
        });
    }

    public static String getMimeType(String url) {
        String type = null;
        String extension = MimeTypeMap.getFileExtensionFromUrl(url);
        if (extension != null) {
            type = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
        }
        return type;
    }
}

File: app/src/main/java/com/example/cooplas/activities/EditProfileScreen.kt
package com.example.cooplas.activities

import android.annotation.TargetApi
import android.app.Activity
import android.app.DatePickerDialog
import android.content.ContentUris
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.database.Cursor
import android.net.Uri
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Environment
import android.provider.DocumentsContract
import android.provider.MediaStore
import android.text.Editable
import android.view.Gravity
import android.view.LayoutInflater
import android.widget.Toast
import com.bumptech.glide.Glide
import com.example.cooplas.R
import com.example.cooplas.models.GeneralRes
import com.example.cooplas.utils.AppConstants
import com.example.cooplas.utils.AppManager
import com.example.cooplas.utils.MarshMallowPermission
import com.jobesk.gong.utils.checkConnection
import com.jobesk.gong.utils.getAccessToken
import com.kaopiz.kprogresshud.KProgressHUD
import com.orhanobut.dialogplus.DialogPlus
import com.orhanobut.dialogplus.OnClickListener
import com.orhanobut.dialogplus.ViewHolder
import kotlinx.android.synthetic.main.activity_edit_profile_screen.*
import kotlinx.android.synthetic.main.activity_edit_profile_screen.et_user_name
import kotlinx.android.synthetic.main.activity_edit_profile_screen.iv_camera
import kotlinx.android.synthetic.main.activity_edit_profile_screen.iv_profile
import kotlinx.android.synthetic.main.activity_edit_profile_screen.rl_female
import kotlinx.android.synthetic.main.activity_edit_profile_screen.rl_male
import kotlinx.android.synthetic.main.activity_edit_profile_screen.tv_female
import kotlinx.android.synthetic.main.activity_edit_profile_screen.tv_male
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.File
import java.util.*

class EditProfileScreen : AppCompatActivity() {
    var gender=""
    var camerafile: File?=null
    var imagefile: File?=null
    lateinit var marshMallowPermission: MarshMallowPermission
    var CAMERA_CODE=1
    var GALLERY_CODE=2

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_edit_profile_screen)
        marshMallowPermission=MarshMallowPermission(this)
        //////////////////Load old Data ////////////////
        Glide.with(this).load(intent.getStringExtra(AppConstants.IMAGE_PATH)?:"").into(iv_profile)
        when(intent.getStringExtra(AppConstants.GENDER)){
            "male"-> updategender(female = false, male = true)
            "female"-> updategender(female = true, male = false)
        }
        et_first_name.text=Editable.Factory.getInstance().newEditable(intent.getStringExtra(AppConstants.FIRST_NAME)?:"")
        et_last_name.text=Editable.Factory.getInstance().newEditable(intent.getStringExtra(AppConstants.LAST_NAME)?:"")
        et_user_name.text=Editable.Factory.getInstance().newEditable(intent.getStringExtra(AppConstants.USERNAME)?:"")

                      ////////////////
        tv_female.setOnClickListener {
            updategender(female = true, male = false)
        }
        tv_male.setOnClickListener {
            updategender(female = false, male = true)
        }
        iv_camera.setOnClickListener {
            showImageSelectionDialog()
        }
        rl_save.setOnClickListener {
            when (checkConnection(this)) {
                true->{ update_profile(et_first_name.text.toString().trim(), et_last_name.text.trim().toString(), et_user_name.text.toString().trim())}
                else->{ Toast.makeText(this,"No Internet Connection", Toast.LENGTH_SHORT).show()}
            }
        }

    }

    private fun updategender(female: Boolean, male: Boolean) {
        when{
            female ->{
                gender="female"
                rl_female.background = resources.getDrawable(R.drawable.round_orange)
                rl_male.background = resources.getDrawable(R.drawable.round_grey)
                tv_female.setTextColor(resources.getColor(R.color.white))
                tv_male.setTextColor(resources.getColor(R.color.black))
            }
            male->{
                gender="male"
                rl_female.background = resources.getDrawable(R.drawable.round_grey)
                rl_male.background = resources.getDrawable(R.drawable.round_orange)
                tv_female.setTextColor(resources.getColor(R.color.black))
                tv_male.setTextColor(resources.getColor(R.color.white))
            }
        }
    }

    private fun showImageSelectionDialog() {
        val dialogeplus: DialogPlus = DialogPlus.newDialog(this)
            .setContentHolder(ViewHolder(LayoutInflater.from(this).inflate(R.layout.select_image_dialog,null,false)))
            .setBackgroundColorResId(android.graphics.Color.TRANSPARENT)
            .setCancelable(false)
            .setGravity(Gravity.BOTTOM)
            .setOnClickListener(uploadImageClickListener)
            .create()
        dialogeplus.show()

    }

    private var uploadImageClickListener: OnClickListener = OnClickListener { dialog, view ->
        when (view.id) {
            R.id.tv_take -> {

                // On Take Photo
                if (!marshMallowPermission.checkPermissionForCamera()) {
                    marshMallowPermission.requestPermissionForCamera()
                } else {
                    openCamera()
                }
                dialog.dismiss()
            }
            R.id.tv_choose -> {

                // On Pick From Gallery
                if (!marshMallowPermission.checkPermissionForExternalStorage()) {
                    marshMallowPermission.requestPermissionForExternalStorage()
                } else {
                    openGallery()
                }
                dialog.dismiss()
            }
            R.id.tv_cancel -> dialog.dismiss()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        if (requestCode == 3) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                openCamera()
            }
        }
        if (requestCode == 2) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                openGallery()
            }
        }
    }

    private fun openCamera() {
        val intent= Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        intent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
        camerafile = File(externalCacheDir,
            System.currentTimeMillis().toString() + ".jpg")
        intent.putExtra(MediaStore.EXTRA_OUTPUT, Uri.fromFile(camerafile))
        startActivityForResult(intent, CAMERA_CODE)
    }

    private fun openGallery() {
        val intent= Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
        startActivityForResult(intent,GALLERY_CODE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == CAMERA_CODE && resultCode == Activity.RESULT_OK) {
            imagefile=camerafile
        }
        if (requestCode == GALLERY_CODE && resultCode == Activity.RESULT_OK) {
            val imageuri = data?.data!!
            imagefile=File(getPath(this,imageuri))
        }
        Glide.with(this).load(imagefile).placeholder(R.drawable.profile_icon).into(iv_profile)
    }


    private fun update_profile(first_name: String, last_name: String, user_name: String) {
        var image : MultipartBody.Part? = null
        if (imagefile==null){
            image=null
        }else{
            image= MultipartBody.Part.createFormData("profile_pic", imagefile?.name, RequestBody.create(MediaType.parse("multipart/form-data"),imagefile))
        }
        val gender = RequestBody.create(MediaType.parse("text/plain"), gender)
        val f_name = RequestBody.create(MediaType.parse("text/plain"), first_name)
        val l_name = RequestBody.create(MediaType.parse("text/plain"), last_name)
        val u_name = RequestBody.create(MediaType.parse("text/plain"), user_name)
//        val image= MultipartBody.Part.createFormData("profile_pic", imagefile?.name,
//            RequestBody.create(MediaType.parse("multipart/form-data"),imagefile))

        val kProgressHUD = KProgressHUD.create(this).show()
        AppManager.getInstance().restClient.cooplas.update_profile("Bearer " + getAccessToken(this), image, gender, f_name, l_name, u_name).enqueue(object :
            Callback<GeneralRes> {

            override fun onFailure(call: Call<GeneralRes>, t: Throwable) {
                kProgressHUD.dismiss()
                Toast.makeText(this@EditProfileScreen, t.message, Toast.LENGTH_SHORT).show()
            }

            override fun onResponse(call: Call<GeneralRes>, response: Response<GeneralRes>) {
                kProgressHUD.dismiss()
                if (response.isSuccessful) {
                    Toast.makeText(this@EditProfileScreen, response.body()?.message, Toast.LENGTH_SHORT).show()
                    finish()
                } else {
                    val obj = JSONObject(response.errorBody()?.string())
                    Toast.makeText(this@EditProfileScreen, obj.get("message").toString(), Toast.LENGTH_SHORT).show()
                }

            }
        })
    }

    @TargetApi(Build.VERSION_CODES.KITKAT)
    fun getPath(context: Context, uri: Uri): String? {

        val isKitKat = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT

        // DocumentProvider
        if (isKitKat && DocumentsContract.isDocumentUri(context, uri)) {
            // ExternalStorageProvider
            if (isExternalStorageDocument(uri)) {
                val docId = DocumentsContract.getDocumentId(uri)
                val split = docId.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val type = split[0]

                if ("primary".equals(type, ignoreCase = true)) {
                    return Environment.getExternalStorageDirectory().toString() + "/" + split[1]
                }

                // TODO handle non-primary volumes
            } else if (isDownloadsDocument(uri)) {

                val id = DocumentsContract.getDocumentId(uri)
                val contentUri = ContentUris.withAppendedId(
                    Uri.parse("content://downloads/public_downloads"), java.lang.Long.valueOf(id))

                return getDataColumn(context, contentUri, null, null)
            } else if (isMediaDocument(uri)) {
                val docId = DocumentsContract.getDocumentId(uri)
                val split = docId.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val type = split[0]

                var contentUri: Uri? = null
                if ("image" == type) {
                    contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI
                } else if ("video" == type) {
                    contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI
                } else if ("audio" == type) {
                    contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
                }

                val selection = "_id=?"
                val selectionArgs = arrayOf(split[1])

                return getDataColumn(context, contentUri, selection, selectionArgs)
            }// MediaProvider
            // DownloadsProvider
        } else if ("content".equals(uri.scheme, ignoreCase = true)) {
            return getDataColumn(context, uri, null, null)
        } else if ("file".equals(uri.scheme, ignoreCase = true)) {
            return uri.path
        }// File
        // MediaStore (and general)

        return null
    }
    /**
     * Get the value of the data column for this Uri. This is useful for
     * MediaStore Uris, and other file-based ContentProviders.
     *
     * @param context       The context.
     * @param uri           The Uri to query.
     * @param selection     (Optional) Filter used in the query.
     * @param selectionArgs (Optional) Selection arguments used in the query.
     * @return The value of the _data column, which is typically a file path.
     */
    fun getDataColumn(context: Context, uri: Uri?, selection: String?, selectionArgs: Array<String>?): String? {

        var cursor: Cursor? = null
        val column = "_data"
        val projection = arrayOf(column)

        try {
            cursor = context.contentResolver.query(uri!!, projection, selection, selectionArgs, null)
            if (cursor != null && cursor.moveToFirst()) {
                val column_index = cursor.getColumnIndexOrThrow(column)
                return cursor.getString(column_index)
            }
        } finally {
            cursor?.close()
        }
        return null
    }
    /**
     * @param uri The Uri to check.
     * @return Whether the Uri authority is ExternalStorageProvider.
     */
    fun isExternalStorageDocument(uri: Uri): Boolean {
        return "com.android.externalstorage.documents" == uri.authority
    }
    /**
     * @param uri The Uri to check.
     * @return Whether the Uri authority is DownloadsProvider.
     */
    fun isDownloadsDocument(uri: Uri): Boolean {
        return "com.android.providers.downloads.documents" == uri.authority
    }
    /**
     * @param uri The Uri to check.
     * @return Whether the Uri authority is MediaProvider.
     */
    fun isMediaDocument(uri: Uri): Boolean {
        return "com.android.providers.media.documents" == uri.authority
    }

}